abba/json/src/Parse.cpp

```diff
@@ -0,0 +1,480 @@
+				{
+//					std::cout << "parsing array" << std::endl;
+					node->type = Json::ARRAY;
+					type = getToken();
+					if (type.type == CURLY_OPEN)
+					{
+						parseTest(node);
+//						node->values.object.emplace(value, node);
+//						while (type.type != ARRAY_CLOSE)
+//						{
+//							parseArray(node);
+//							_file->seekg(_prevPos);
+//							type = getToken();
+//							if (type.type == COMMA)
+//								type = getToken();
+//						}
+					}
+					else
+					{
+						_file->seekg(_prevPos);
+						node->values.str = value;
+						node->values.list.push_back(node);
+						parseArray1(node);
+					}
+					break;
+				}
+				case ARRAY_CLOSE:
+				{
+					endObj = true;
+					break;
+				}
+				case COMMA:
+					break;
+				case NULL_TYPE:
+				{
+					node->type = Json::NULL_TYPE;
+					node->values.str = type.value;
+					node->values.object.emplace(value, node);
+					break;
+				}
+			}
+			if (getToken().type == CURLY_CLOSE)
+			{
+//				endObj = true;
+			}
+		}
+	}
+	return node;
+}
+
+void Parse::parseTest(Json *node)
+{
+	bool	finish = false;
+	while (!finish)
+	{
+		Type type = getToken();
+		std::string value = type.value;
+		if (getToken().type != COLON)
+		{
+			std::cout << "bus error" << std::endl;
+			exit(0);
+		}
+		type = getToken();
+		switch (type.type)
+		{
+			case CURLY_OPEN:
+				break;
+			case CURLY_CLOSE:
+				break;
+			case STRING:
+			{
+				node->type = Json::STRING;
+				node->values.str = type.value;
+				std::cout << type.value << std::endl;
+				node->values.object.emplace(value, node);
+				break;
+			}
+			case NUMBER:
+			{
+				node->type = Json::NUMBER;
+				node->values.number = std::stoi(type.value);
+				node->values.object.emplace(value, node);
+				break;
+			}
+			case BOOLEAN:
+			{
+				node->type = Json::BOOLEAN;
+				(type.value == "true") ? node->values.boolean = true : node->values.boolean = false;
+				node->values.object.emplace(value, node);
+				break;
+			}
+			case COLON:
+				break;
+			case ARRAY_OPEN:
+			{
+				node->type = Json::ARRAY;
+//				type = getToken();
+				if (type.type == CURLY_OPEN)
+				{
+					node->values.object.emplace(value, node);
+					while (type.type != ARRAY_CLOSE)
+					{
+						parseArray(node);
+						_file->seekg(_prevPos);
+						type = getToken();
+						if (type.type == COMMA)
+							type = getToken();
+					}
+				}
+				else
+				{
+//					_file->seekg(_prevPos);
+					_prevPos = _file->tellg();
+					node->values.str = value;
+					node->values.list.push_back(node);
+					parseArray1(node);
+					_file->seekg(_prevPos);
+					finish = true;
+				}
+				break;
+			}
+			case ARRAY_CLOSE:
+				break;
+			case COMMA:
+				break;
+			case NULL_TYPE:
+			{
+				node->type = Json::NULL_TYPE;
+				node->values.str = type.value;
+				node->values.object.emplace(value, node);
+				break;
+			}
+		}
+		if (getToken().type == CURLY_CLOSE)
+		{
+			finish = true;
+			_prevPos = _file->tellg();
+		}
+	}
+}
+
+void Parse::parseArray(Json *node)
```

Comment:
Ideally, a function like this would be called when you encounter the array open token `[` and it would parse the entire array until the `]` token, returning only when it is finished. On encountering any more opening tokens like `[` and `{` you could parse those recursively using the same function logic.